Add support for implementing arbitrary encoders

Milyardo opened the pull request:
This pull request adds the ability for a user to specify their own arbitrary encoder in `XMLRPCRequest`. It creates a protocol for encoders to implement and adds a `-(void) setEncoder: *id<XMLRPCEncoder> encoder` method to `XMLRPCRequest`.

These commits should add that support without changing the any behavior of `XMLRPC`.

Commit:
Refactor XMLRPCEncoder to XMLRPCEncoderImpl.

This change renames XMLRPCEncoder to XMLRPCEncoderImpl to imply that
XMLRPCEncoder is actually an implentation of a interface yet to be defined. The
yet to be defined interface will allow users of the xmlrpc to prvide theier own
encoders to support XMLRPC extensions.

Commit:
Create a protocol for XMLRPCEncoderImpl to implement.

This protocol will be used instead of XMLRPCEncoderImpl directly.

Commit:
Have XMLRPC Request use the new XMLRPCEncoder protocol instead of XMLRPCEncoderImpl

Commit:
Add a setEncoder method.

Commit:
Remove header created by XCode

eczarny commented:
I guess I'm still not sure how helpful it would be for people to provide their own encoders. What is the use case you have for implementing this?

Milyardo commented:
I outline the case for implementing arbitrary encoders here:

https://github.com/eczarny/xmlrpc/issues/14

To summarize some xmlrpc servers implement non-standard extensions to the XMLRPC specification. For example the Apache XMLRPC library implements 64 bit integers using `<ex:i8>` tags, while other servers may just call them `<i8>`. Giving users the ability to specify arbitrary encoders will allow users add support to alternative dialects.

Outside custom datatypes other extensions to the XMLRPC spec include streaming, compression, server-side introspection, concurrnecy, and exception handling. A custom encoder an parser can be used to programmatically give users the ability provide implementations for these extensions.

Milyardo commented:
I would like to reiterate this implementation changes at little behavior as possible. For example, ideally  in a more formal implementation several currently private methods like `(NSString *)encodeObject:` and `(NSString *)valueTag: would have to moved from`XMLRPCEncoderImpl`'s private header to the protocol.

It would also make more sense to provide encoders and parsers to the `XMLRPCConnectionManager` and have the connection manager produce Request/Response objects with the proper encoder set.

eczarny commented:
Makes sense to me. I'll go through your changes this afternoon. Thanks!

eczarny commented on XMLRPCEncoderImpl.h:
```diff
@@ -0,0 +1,21 @@
+#import <Foundation/Foundation.h>
+#import "XMLRPCEncoder.h"
+
+@interface XMLRPCEncoderImpl : NSObject <XMLRPCEncoder> {
```
Could we rename this to XMLRPCDefaultEncoder? I never liked the *Impl naming convention.

Milyardo replied:
I chose the *Impl, naming because of my Java background but from a quick google search, I think in Objective-C calling them delegates are more appropriate? However, `XMLRPCDefaultEncoderDelegate` sounds terrible and `XMLRPCDefaultEncoder` sounds great to me.

eczarny commented on XMLRPCRequest.h:
```diff
@@ -1,10 +1,10 @@
 @interface XMLRPCRequest : NSObject {
     NSMutableURLRequest *myRequest;
-    XMLRPCEncoder *myXMLEncoder;
+    id <XMLRPCEncoder> myXMLEncoder;
```
No space between `id` and `<XMLRPCEncoder>`.

eczarny commented on XMLRPCRequest.h:
```diff
@@ -22,6 +22,7 @@
 - (NSString *)userAgent;
 
 #pragma mark -
+- (void)setEncoder: (id <XMLRPCEncoder>) encoder;
```
Same here, no space between `id` and `<XMLRPCEncoder>`.

eczarny commented on XMLRPCRequest.m:
```diff
@@ -12,7 +13,7 @@ - (id)initWithURL: (NSURL *)URL {
         }
         
-        myXMLEncoder = [[XMLRPCEncoder alloc] init];
+        myXMLEncoder = [[XMLRPCEncoderImpl alloc] init];
```
I think it would make sense to specify an encoder when creating the request. Perhaps we could add a new constructor?

``` objc
- (id)initWithURL: (NSURL *)URL encoder: (id<XMLRPCEncoder>)encoder;
```

This would then become the designated constructured, `initiWithURL:` could just call this new constructor with the default XMLRPCEncoder implementation.

Milyardo replied:
I agree, I considered that exact implementation, but I didn't want to change the break the existing API without feedback.

Milyardo commented on XMLRPCEncoderImpl.h:
```diff
@@ -0,0 +1,21 @@
+- (NSString *)method;
+
+- (NSArray *)parameters;
+
```
It is my understanding that it is unnecessary to re-declare prototypes that are already defined in a protocol. I'll remove lines 5-19 as they are redundant.

Milyardo commented on XMLRPCRequest.m:
```diff
@@ -44,6 +45,15 @@ - (NSString *)userAgent {
+    //Copy the old method and parameters to the new encoder.
+    NSString *method = [myXMLEncoder method];
+    NSArray *parameters = [myXMLEncoder parameters];
+    myXMLEncoder = encoder;
```
The note above is referencing this line here. I don't know how to handle assumptions about ownership of references in objective-c without ARC. Should we assume if a user provided their own encoder that they're responsible for releasing it?

The previous encoder could been instantiated with the object, or could have been user created.

eczarny replied:
I believe that the XMLRPCRequest should take ownership of the encoder. It would probably be wise to retain the new encoder and release the previous encoder. There should be some decent example of this pattern elsewhere in the framework.

Commit:
Refactor XMLRPCDencoderImpl to XMLRPCDefaultEndcoder

This refactor was made in response to a discussion on naming conventions here:
https://github.com/eczarny/xmlrpc/pull/17#r645514

Commit:
Remove unnecessary prototypes from XMLRPCDefaultEncoder

The methods in defined in XMLRPCDefaultEncoder.h are already defined in the
XMLRPCEncoder protocol and there are are redundant.

Commit:
Add a constuctor to initialize XMLRPCRequest with a given encoder.

Commit:
HAve XMLRPCREquest take ownership of encoders assigned to it.

This commit causes XMLRPCRequest to retain all XMLRPCEncoders assigned to it.

Milyardo commented:
Pushed new commits addressing all concerns.

eczarny commented on XMLRPCRequest.m:
```diff
@@ -12,12 +13,18 @@ - (id)initWithURL: (NSURL *)URL {
 }
 
+- (id)initWithURL: (NSURL *)URL {
+    XMLRPCDefaultEncoder *encoder = [XMLRPCDefaultEncoder alloc];
```
Should there be an `init` call here as well? This constructor would probably be better written as:

``` objc
- (id)initWithURL: (NSURL *)URL {
    return [self initWithURL: URL withEncoder: [[XMLRPCDefaultEncoder alloc] init]];
}
```

Milyardo replied:
You are correct. I did not realize there was was a semantic difference between alloc, and the default constructor.

Milyardo commented:
How well does github handle operations that destructive to the history with regards to pull request? I plan on squashing these  revisions after your approval to one commit. Will I have rebase them in a new branch and open a new pull request on that?

Commit:
Call default constructor when initailizing the default XMLRPC Encoder

eczarny commented:
I'm not sure hot GitHub would handle the rebase. It could handle it gracefully. Give it a shot, if it doesn't work I'll close this pull request and you can create another.
